apps/shop/tests.py

```diff
@@ -6,34 +9,168 @@
+
+    def get_detail_url(self, _id):
+        return reverse(f"{self.basename}-detail", args=(_id,))
+
+    def _add_saldo_to_user(self, amount: int):
+        return G(
+            PaymentTransaction,
+            user=self.user,
+            amount=amount,
+            source=TransactionSource.CASH,
+        )
+
+    def _perform_purchase(self, order_data: List[dict]):
+        data = {"user": self.user.id, "orders": order_data}
+        return self.client.post(
+            self.get_list_url(), data=data, **self.generate_headers()
+        )
+
+    def test_insufficient_funds(self):
+        item: Item = G(Item, available=True, price=100)
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
+        self.assertEqual(response.json().get("detail"), "Insufficient funds")
+
+    def test_purchase_order_line(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_purchase_not_available(self):
+        item: Item = G(Item, available=False, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_saldo_corresponds_with_purchase_amount(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+        self.assertEqual(self.user.saldo, item.price)
+
+        self._perform_purchase([{"object_id": item.id, "quantity": 1}])
+
+        self.assertEqual(self.user.saldo, 0)
+
+    def test_purchase_requires_an_item_quantity(self):
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price)
+
+        response = self._perform_purchase([{"object_id": item.id}])
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_purchase_quantity(self):
+        quantity = 3
+        item: Item = G(Item, available=True, price=100)
+        self._add_saldo_to_user(item.price * quantity)
+
+        response = self._perform_purchase(
+            [{"object_id": item.id, "quantity": quantity}]
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(self.user.saldo, 0)
+
+    def test_purchase_quantity_requires_saldo_for_all(self):
+        price = 100
+        quantity = 3
+        item: Item = G(Item, available=True, price=price)
+        self._add_saldo_to_user(price)
+
+        response = self._perform_purchase(
+            [{"object_id": item.id, "quantity": quantity}]
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_406_NOT_ACCEPTABLE)
```

Comment:
Okay so this was perhaps not added with this PR, but I actually highly disagree that this should be a `406 Not Acceptable`-request. That status code is used for somthing completely [else](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406). If we want to use a status code to check this, I suggest we use `402 Request Failed / Payment Required` in accordance with [Stripe's standards](https://stripe.com/docs/api/errors) , while not _officially_ standard, I think this is an appropiate usecase

Reply:
That may be, but I chose not to change the API, since that would be a breaking change.